src/main/java/gregtech/api/metatileentity/multiblock/ui/MultiblockUIBuilder.java

```diff
@@ -0,0 +1,982 @@
+                        "gregtech.multiblock.universal.problem.crowbar"));
+            }
+        }
+        return this;
+    }
+
+    /**
+     * Adds two error lines when the machine's muffler hatch is obstructed.
+     * <br>
+     * Added if the structure is formed and if the passed parameter is true.
+     */
+    public MultiblockUIBuilder addMufflerObstructedLine(boolean isObstructed) {
+        if (!isStructureFormed) return this;
+        if (getSyncer().syncBoolean(isObstructed)) {
+            addKey(KeyUtil.lang(TextFormatting.RED,
+                    "gregtech.multiblock.universal.muffler_obstructed"));
+            addKey(KeyUtil.lang(TextFormatting.GRAY,
+                    "gregtech.multiblock.universal.muffler_obstructed_desc"));
+        }
+        return this;
+    }
+
+    /**
+     * Adds a fuel consumption line showing the fuel name and the number of ticks per recipe run.
+     * <br>
+     * Added if structure is formed, the machine is active, and the passed fuelName parameter is not null.
+     */
+    public MultiblockUIBuilder addFuelNeededLine(String fuelName, int previousRecipeDuration) {
+        if (!isStructureFormed || !isActive) return this;
+        fuelName = getSyncer().syncString(fuelName);
+        previousRecipeDuration = getSyncer().syncInt(previousRecipeDuration);
+        addKey(KeyUtil.lang(TextFormatting.GRAY,
+                "gregtech.multiblock.turbine.fuel_needed",
+                KeyUtil.string(TextFormatting.RED, fuelName),
+                KeyUtil.number(TextFormatting.AQUA, previousRecipeDuration)));
+        return this;
+    }
+
+    /**
+     * Adds the name of a recipe map to the display.
+     *
+     * @param map the {@link RecipeMap} to get the name of
+     */
+    public MultiblockUIBuilder addRecipeMapLine(RecipeMap<?> map) {
+        if (!isStructureFormed) return this;
+
+        IKey mapName = KeyUtil.lang(TextFormatting.YELLOW, map.getTranslationKey());
+        addKey(KeyUtil.lang(TextFormatting.GRAY, "gregtech.multiblock.machine_mode", mapName));
+
+        return this;
+    }
+
+    /**
+     * Adds the current outputs of a recipe from recipe logic. Items then fluids.
+     *
+     * @param arl an instance of an {@link AbstractRecipeLogic} to gather the outputs from.
+     */
+    public MultiblockUIBuilder addRecipeOutputLine(@NotNull AbstractRecipeLogic arl) {
+        return addRecipeOutputLine(arl, DEFAULT_MAX_RECIPE_LINES);
+    }
+
+    /**
+     * Adds the current outputs of a recipe from recipe logic. Items then fluids.
+     *
+     * @param arl      an instance of an {@link AbstractRecipeLogic} to gather the outputs from.
+     * @param maxLines the maximum number of lines to print until truncating with {@code ...}
+     */
+    public MultiblockUIBuilder addRecipeOutputLine(AbstractRecipeLogic arl, int maxLines) {
+        // todo doing this every tick on the server is probably not good
+        Recipe recipe = arl.getParallelRecipe();
```

Comment:
Since this is for multiblocks, this needs to be calling the multiblock specific function that takes into account distinct buses, as in MultiblockRecipeLogic#prepareRecipeDistinct().

Is there even a need for this though, if we can return the recipe from ARL, and then use the field `parallelRecipesPerformed`?